Overhaul to how encasing blocks works

Cyvack opened the pull request:
Changes the way that encasing shafts / cogwheels / any other number of custom blocks works

     - Adds 2 new interfaces, IEncasable and IEncased. These interfaces are javadoc'd, So using them shouldn't be too difficult. 
     - -IEncasable essentially indicates that the block is able to be encased, and gives a method that is used when tryEncase is successful. 
     - -The IEncased interface indicates that this block is the result of encasing another block. It adds methods to grab the casing / block required to make the encased version.
     -Adds an Encasable class, to add a hashmap of all of the IEncasable blocks, and an ArrayList of valid IEncased to go with each IEncasable key. 
     - -This encasable class is called when making a new IEncased block, via an .onRegisteredAfter event in the block's registry.

Cyvack committed:
Make Encasing utilize Interfaces

Cyvack committed:
Add Javadoc to IEncasable and IEncased interfaces

Cyvack committed:
Fix Weird block.block issues (No clue where those came from)

Cyvack committed:
Final Touch ups for now and made requirement for casing normal Block instead of CasingBlock

PepperCode1 commented on src/main/java/com/simibubi/create/AllBlocks.java:
```diff
@@ -359,50 +360,56 @@ public class AllBlocks {
 			.register();
 
 	public static final BlockEntry<EncasedShaftBlock> ANDESITE_ENCASED_SHAFT =
-		REGISTRATE.block("andesite_encased_shaft", EncasedShaftBlock::andesite)
```
Don't change the indentation level

PepperCode1 commented on src/main/java/com/simibubi/create/AllBlocks.java:
```diff
@@ -359,50 +360,56 @@ public class AllBlocks {
-			.transform(BuilderTransformers.encasedShaft("brass", () -> AllSpriteShifts.BRASS_CASING))
-			.transform(axeOrPickaxe())
-			.register();
+			REGISTRATE.block("brass_encased_shaft", EncasedShaftBlock::new)
```
Don't change the indentation level

PepperCode1 commented on src/main/java/com/simibubi/create/content/contraptions/relays/elementary/IEncasable.java:
```diff
@@ -0,0 +1,47 @@
+	default InteractionResult tryEncase(BlockState state, Level level, BlockPos pos,ItemStack heldItem) {
+		ArrayList<Block> encasedBlocks = Encasable.encasableBlocks.get(state.getBlock());
+		for (Block block : encasedBlocks) {
+			if(block instanceof IEncased encased){
```
Missing spaces before and after parentheses

PepperCode1 commented on src/main/java/com/simibubi/create/content/contraptions/relays/elementary/IEncased.java:
```diff
@@ -0,0 +1,11 @@
+ * Implement this interface to indicate that this block is an EncasedBlock
+ */
+public interface IEncased {
+	Block getCasing();
```
Missing new line between methods

PepperCode1 commented on src/main/java/com/simibubi/create/content/contraptions/relays/elementary/IEncased.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Implement this interface to indicate that this block is an EncasedBlock
+ */
+public interface IEncased {
```
The I- prefix is legacy and should not be used on new interfaces. This should be renamed to `Encased`.

PepperCode1 commented on src/main/java/com/simibubi/create/content/contraptions/relays/encased/EncasedShaftBlock.java:
```diff
@@ -54,7 +46,7 @@ public InteractionResult onSneakWrenched(BlockState state, UseOnContext context)
 			return InteractionResult.SUCCESS;
 		context.getLevel()
-			.levelEvent(2001, context.getClickedPos(), Block.getId(state));
+			.levelEvent(2001, context.getClickedPos(), ShaftBlock.getId(state));
```
This is a static method. Don't change `Block` to `ShaftBlock`.

PepperCode1 commented on src/main/java/com/simibubi/create/foundation/data/Encasable.java:
```diff
@@ -0,0 +1,26 @@
+public class Encasable {
+	public static Map<Block, ArrayList<Block>> encasableBlocks = new HashMap<>();
+
+	public static void register(IEncased encasedBlock, IEncasable blockToBeEncased, CasingBlock casing){
```
- Missing space after right parenthesis
- All of this method's code has an extra level of indentation

PepperCode1 commented on src/main/java/com/simibubi/create/foundation/data/Encasable.java:
```diff
@@ -0,0 +1,26 @@
+package com.simibubi.create.foundation.data;
```
Move to `com.simibubi.create.content.contraptions.relays.elementary`

PepperCode1 commented on src/main/java/com/simibubi/create/foundation/data/Encasable.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Encasable {
```
Rename to `EncasableRegistry`

PepperCode1 commented on src/main/java/com/simibubi/create/content/contraptions/relays/elementary/CogWheelBlock.java:
```diff
@@ -122,38 +122,8 @@ public InteractionResult use(BlockState state, Level world, BlockPos pos, Player
-			}
-
-			KineticTileEntity.switchToBlockState(world, pos, encasedState);
+		if (!tryEncase(state, world, pos, heldItem).equals(InteractionResult.PASS))
```
Use `InteractionResult#consumesAction()` here

PepperCode1 commented on src/main/java/com/simibubi/create/content/contraptions/fluids/pipes/FluidPipeBlock.java:
```diff
@@ -114,17 +115,12 @@ public void setPlacedBy(Level pLevel, BlockPos pPos, BlockState pState, LivingEn
-		if (world.isClientSide)
+
+		ItemStack heldItem = player.getItemInHand(hand);
+		if (!tryEncase(state, world, pos, heldItem).equals(InteractionResult.PASS))
```
Use `InteractionResult#consumesAction()` here

PepperCode1 commented on src/main/java/com/simibubi/create/content/contraptions/relays/elementary/IEncasable.java:
```diff
@@ -0,0 +1,47 @@
+	 * @return If the Interaction result was a success, or pass
+	 */
+	default InteractionResult tryEncase(BlockState state, Level level, BlockPos pos,ItemStack heldItem) {
+		ArrayList<Block> encasedBlocks = Encasable.encasableBlocks.get(state.getBlock());
```
- `EncasableRegistry` should have a static method to run `encasableBlocks.get(block)` so that the field is not accessed directly here
- Use `List` instead of `ArrayList`

PepperCode1 commented on src/main/java/com/simibubi/create/foundation/data/Encasable.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Map;
+
+public class Encasable {
+	public static Map<Block, ArrayList<Block>> encasableBlocks = new HashMap<>();
```
Use `List` instead of `ArrayList` here

PepperCode1 commented on src/main/java/com/simibubi/create/content/contraptions/relays/encased/EncasedShaftBlock.java:
```diff
@@ -25,26 +25,18 @@
+	public Block getCasing() {
 		return casing;
 	}
+	public void setCasing(Block casing) {this.casing = casing;}
```
- Missing a new line before this method
- Missing new line after left bracket
- Missing new line before right bracket

Cyvack commented:
Alright, thanks for pointing out these changes
I'll make them and do a push

Cyvack committed:
Make requested changes

Cyvack committed:
Add more parameters for tryEncasing, for use in handleEncasing
Move handleEncasing to Encased Interface for more flexability

PepperCode1 commented:
In my first review I only commented on code style, which has mostly been fixed now, but I didn't have time to review the API design. I looked at it carefully now and it doesn't seem quite right. Instances of `Block` should be immutable, but adding a setter to some instances breaks their immutability, so the setter must be removed. There are some other nitpicks that should be resolved. I'll push a commit to this PR soon that will resolve all of this.

Cyvack commented:
I was in the middle of re-writing some of the API to get rid of that setter method, as I realized that it may cause some issues with adding multiple encasement options to the same block
Though if you're already in the process of making a commit to fix those changes and other nitpicks, I can hold off on my changes.

PepperCode1 committed:
Simplify and organize

- Rename:
  - Encasable -> EncasableBlock
  - Encased -> EncasedBlock
  - EncasableRegistry -> EncasingRegistry
- Remove EncasedBlock#setCasing
- Remove encasedBlock argument from EncasedBlock#handleEncasing
- Add Registrate builder transformer to EncasingRegistry for easy use

cakeGit commented:
so it is you who I have to bless for saving me pain in my modding

and then also causing more when I had to downgrade from the dev version
